### PR TITLE
Data, Data specs, and Data processing

### PR DESCRIPTION
**Pull request by mbannick**

### CODE REVIEW

File: src/placeholder/data/data.py [hunk 1]
```diff
@@ -72,6 +90,42 @@ def _validate_specs(self, df):
         for spec in self._data_specs:
             spec._validate_df(df=df)
 
+    @staticmethod
+    def _convert_to_numpy(df: pd.DataFrame, column: Union[str, List[str]]) -> np.ndarray:
+        return np.asarray(df[column])
+
+    def _process_data_with_spec(self, df: pd.DataFrame, spec: DataSpecs):
```
[the remark this answers is not in the document]
Comment (mbannick) [reply src/placeholder/data/data.py:97]: Added `self.data_attributes`: 20a3f4b

File: src/placeholder/data/data_specs.py [hunk 2]
```diff
@@ -22,12 +26,25 @@ class DataSpecCompatibilityError(PlaceholderError):
...
-    col_groups: List[str]
     col_obs_se: str
+    col_groups: List[str] = None
 
     def __post_init__(self):
         pass
 
+    @property
+    def _col_attributes(self):
+        return [x for x in dir(self) if isinstance(x, str)
```
[the remark this answers is not in the document]
Comment (mbannick) [reply src/placeholder/data/data_specs.py]: Yes! This is great: 20a3f4b

File: src/placeholder/data/data.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zhengp0) [reply src/placeholder/data/data.py:97]: Thanks Marlena! If we could name `self.data_attributes` to `self.data_spec_col_attributes` or something else that will be great! Just to be consistent with how we define the `data_attributes` in `DataSpecs` :)

[the remark this answers is not in the document]
Comment (mbannick) [reply src/placeholder/data/data.py:97]: Yes! And I'm working on the second part of this -- better processing now that we have `self.data_spec_col_attributes`. Will update :)

[the remark this answers is not in the document]
Comment (mbannick) [reply src/placeholder/data/data.py:105]: Alright, it is here: de4ad69

**Pull request merged**